lxd/patches.go

```diff
@@ -2504,6 +2505,187 @@ func patchStorageApiCephSizeRemove(name string, d *Daemon) error {
+		}
+
+		onDiskDevices, err := devDir.Readdirnames(-1)
+		if err != nil {
+			logger.Errorf("Failed to read directory entries from \"%s\": %s", devicesPath, err)
+			return err
+		}
+
+		// nothing to do
+		if len(onDiskDevices) == 0 {
+			logger.Debugf("Devices directory \"%s\" is empty", devicesPath)
+			continue
+		}
+
+		hasDeviceEntry := map[string]bool{}
+		for _, v := range onDiskDevices {
+			key := fmt.Sprintf("%s/%s", devicesPath, v)
+			hasDeviceEntry[key] = false
+		}
+
+		// Load the container from the database.
+		c, err := containerLoadByName(d.State(), ct)
+		if err != nil {
+			logger.Errorf("Failed to load container %s: %s", ct, err)
+			return err
+		}
+
+		isRunning := c.IsRunning()
+
+		// go through all devices for each container
+		expandedDevices := c.ExpandedDevices()
+		for _, name := range expandedDevices.DeviceNames() {
+			d := expandedDevices[name]
+
+			// We only care about unix-{char,block} and disk devices
+			// since other devices don't create on-disk files.
+			if d["type"] != "disk" && !shared.StringInSlice(d["type"], []string{"unix-char", "unix-block"}) {
+				continue
+			}
+
+			if d["type"] == "disk" {
+				relativeDestPath := strings.TrimPrefix(d["path"], "/")
+				hyphenatedDevName := strings.Replace(relativeDestPath, "/", "-", -1)
+				devNameLegacy := fmt.Sprintf("disk.%s", hyphenatedDevName)
+				devPathLegacy := filepath.Join(devicesPath, devNameLegacy)
+
+				// Try to unmount disk devices otherwise we get
+				// EBUSY when we try to rename block devices.
+				// But don't error out.
+				syscall.Unmount(devPathLegacy, syscall.MNT_DETACH)
+
+				// wipe stale devices
+				if !isRunning {
+					if !shared.PathExists(devPathLegacy) {
+						logger.Debugf("Device \"%s\" does not exist", devPathLegacy)
+						continue
+					}
+
+					err := os.Remove(devPathLegacy)
+					if err != nil {
+						logger.Errorf("Failed to remove device \"%s\": %s", devPathLegacy, err)
+						return err
+					}
+
+					continue
+				}
+
+				if !shared.PathExists(devPathLegacy) {
+					logger.Debugf("Device \"%s\" does not exist", devPathLegacy)
+					continue
+				}
+
+				hasDeviceEntry[devPathLegacy] = true
+
+				// Switch device to new device naming scheme.
+				devPathNew := filepath.Join(devicesPath, fmt.Sprintf("disk.%s.%s", name, hyphenatedDevName))
+				err = os.Rename(devPathLegacy, devPathNew)
+				if err != nil {
+					logger.Errorf("Failed to rename device from \"%s\" to \"%s\": %s", devPathLegacy, devPathNew, err)
+					return err
+				}
+			}
+
+			srcPath := d["source"]
+			if srcPath == "" {
+				srcPath = d["path"]
+			}
+
+			relativeSrcPathLegacy := strings.TrimPrefix(srcPath, "/")
+			hyphenatedDevNameLegacy := strings.Replace(relativeSrcPathLegacy, "/", "-", -1)
+			devNameLegacy := fmt.Sprintf("unix.%s", hyphenatedDevNameLegacy)
+			devPathLegacy := filepath.Join(devicesPath, devNameLegacy)
+
+			// Wipe stale devices.
+			if !isRunning {
```

Comment:
Wouldn't it be simpler to just have logic that unmount and wipes all files/directories in /var/lib/lxd/devices for any container which isn't running?

Reply:
It just felt more uniform this way.

Reply:
improved